build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const BAIL_OUT_CONDITIONS = {
+  // If this isn't a FunctionDeclaration, bail out on modification.
+  isNotFunction: (t, path) => !t.isFunctionDeclaration(path.node),
+
+  // If this FunctionDeclaration is a Generator, bail out on modification.
+  isGenerator: (t, path) => path.node.generator,
+
+  // If this FunctionDeclaration doesn't have a single return statement, bail out on modification
+  isNotSingleReturnStatment: (t, path) => {
+    if (!path.get('body').isBlockStatement()) {
+      return false;
+    }
+    const body = path.get('body.body.0');
+    return !body || !body.isReturnStatement();
+  },
+
+  // Since we don't know if exported members are 'new'd, bail out on modification.
+  isExported: (t, path) =>
+    path.findParent(
+      path =>
+        path.isExportNamedDeclaration() || path.isExportDefaultDeclaration()
+    ),
+
+  // If the function contains usage of `arguments`, bail out on modification.
+  containsArgumentsUsage: (t, path) => {
+    let containsArgumentsUsage = false;
+    path.traverse({
+      Identifier(path) {
+        if (t.isIdentifier(path.node, {name: 'arguments'})) {
+          containsArgumentsUsage = true;
+          path.stop();
+        }
+      },
+    });
+    return containsArgumentsUsage;
+  },
+
+  // If the function contains a ThisExpression, converting from a FunctionDeclaration to a
+  // VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+  // Bail out on modification.
+  containsThisExpression: (t, path) => {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  },
+
+  // If the FunctionDeclaration identifier is newed in the scope of this program, bail out on modification.
+  isNewedInProgramScope: (t, path) => {
+    const {name} = path.get('id').node;
+    let isNewed = false;
+    path
+      .findParent(path => path.isProgram())
+      .traverse({
+        NewExpression(path) {
+          if (t.isIdentifier(path.node.callee, {name})) {
+            isNewed = true;
+            path.stop();
+          }
+        },
+      });
+    return isNewed;
+  },
+};
+
+function createVariableDeclaration(t, path) {
```

Comment:
Nit: Everything should be inside the `module.exports` function, so that you don't need to keep passing around the `t` variable. The function will only ever be invoked once, so it won't slow anything down.